src/servers/src/tls.rs

```diff
@@ -110,6 +130,104 @@ impl TlsOption {
     pub fn should_force_tls(&self) -> bool {
         !matches!(self.mode, TlsMode::Disable | TlsMode::Prefer)
     }
+
+    pub fn cert_path(&self) -> &Path {
+        Path::new(&self.cert_path)
+    }
+
+    pub fn key_path(&self) -> &Path {
+        Path::new(&self.key_path)
+    }
+}
+
+/// A mutable container for TLS server config
+///
+/// This struct allows dynamic reloading of server certificates and keys
+pub struct ReloadableTlsServerConfig {
+    tls_option: TlsOption,
+    config: RwLock<Option<Arc<ServerConfig>>>,
+    version: AtomicUsize,
+}
+
+impl ReloadableTlsServerConfig {
+    /// Create server config by loading configuration from `TlsOption`
+    pub fn try_new(tls_option: TlsOption) -> Result<ReloadableTlsServerConfig> {
+        let server_config = tls_option.setup()?;
+        Ok(Self {
+            tls_option,
+            config: RwLock::new(server_config.map(Arc::new)),
+            version: AtomicUsize::new(0),
+        })
+    }
+
+    /// Reread server certificates and keys from file system.
+    pub fn reload(&self) -> Result<()> {
+        let server_config = self.tls_option.setup()?;
+        *self.config.write().unwrap() = server_config.map(Arc::new);
+        self.version.fetch_add(1, Ordering::Relaxed);
+        Ok(())
+    }
+
+    /// Get the server config hold by this container
+    pub fn get_server_config(&self) -> Option<Arc<ServerConfig>> {
+        self.config.read().unwrap().clone()
+    }
+
+    /// Get associated `TlsOption`
+    pub fn get_tls_option(&self) -> &TlsOption {
+        &self.tls_option
+    }
+
+    /// Get version of current config
+    ///
+    /// this version will auto increase when server config get reloaded.
+    pub fn get_version(&self) -> usize {
+        self.version.load(Ordering::Relaxed)
+    }
+}
+
+pub fn watch_tls_config(tls_server_config: Arc<ReloadableTlsServerConfig>) -> Result<()> {
+    if tls_server_config.get_tls_option().mode == TlsMode::Disable {
+        return Ok(());
+    }
+
+    let tls_server_config_for_watcher = tls_server_config.clone();
+
+    let (tx, rx) = channel::<notify::Result<notify::Event>>();
+    let mut watcher = notify::recommended_watcher(tx).context(FileWatchSnafu)?;
+
+    watcher
+        .watch(
+            tls_server_config.get_tls_option().cert_path(),
+            RecursiveMode::NonRecursive,
+        )
+        .context(FileWatchSnafu)?;
+
+    watcher
+        .watch(
+            tls_server_config.get_tls_option().key_path(),
+            RecursiveMode::NonRecursive,
+        )
+        .context(FileWatchSnafu)?;
+
+    std::thread::spawn(move || {
+        let _watcher = watcher;
```

Comment:
question: if we need to keep `watcher` in the thread, why not put the creation in the thread? It seems more straightforward.

Reply:
because the creation of watcher may raise various error. If we put watcher creation in the thread, we can only log them instead of returning to caller.